### PR TITLE
fix: rewrite handling of missing keys for indexers

### PR DESCRIPTION
**Pull request by cristianmtr**
This will be required for the CRUD integration tests suite

- handling of missing keys should not `Raise`
- missing keys are filtered out of the request
- we do log a `warning` - up to the user to interpret if that's expected or not

### CODE REVIEW

File: jina/executors/indexers/keyvalue.py [hunk 1]
```diff
@@ -78,22 +79,17 @@ def query(self, key: int) -> Optional[bytes]:
...
-        self.query_handler.close()
-        self.handler_mutex = False
-        self.delete(keys)
+        keys = check_keys_exist(keys, self.query_handler.header.keys(), self.save_abspath, self.logger)
+        self.delete(keys, **{'keys_precomputed': True})
         self.add(keys, values)
         return
 
     def delete(self, keys: Iterator[int], *args, **kwargs):
+        if kwargs.get('keys_precomputed') is not True:
```
Comment (JoanFM) [inline jina/executors/indexers/keyvalue.py:88]: why do we pass extra arguments? not expected in such key interface I would say

File: jina/helper.py [hunk 2]
```diff
@@ -693,12 +693,16 @@ def run(self):
         return asyncio.run(func(*args, **kwargs))
 
 
-def check_keys_exist(keys_to_check, existing_keys):
-    missed = []
-    for k in keys_to_check:
-        if k not in existing_keys:
-            missed.append(k)
-    return missed
+def check_keys_exist(keys: Iterator, keys_existing: Iterator, save_abspath: str, logger: 'JinaLogger'):
```
Comment (JoanFM) [inline jina/helper.py:699]: the name save abspath of this argument links it heavily with an executor, I would call it check_path

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/helper.py (hunk 2 above)
Comment (cristianmtr) [reply jina/helper.py]: Done

File: jina/executors/indexers/keyvalue.py (hunk 1 above)
Comment (JoanFM) [reply jina/executors/indexers/keyvalue.py:88]: the interfaces should never be changed, pasa executor parameter if needed but do not change interface or assume things will be passed

Comment (cristianmtr) [reply jina/executors/indexers/keyvalue.py:88]: It's a small optimization. 

When we call an `.update`, the keys will be computed for the update anyway. The update will then call the `delete` with only the filtered out keys. 

If we don't do it this way, the `.delete` method will have to recalculate and filter out the missing keys.

I can add a comment

File: jina/executors/indexers/keyvalue.py [hunk 3]
```diff
@@ -78,22 +79,17 @@ def query(self, key: int) -> Optional[bytes]:
...
-            if self.query_handler.header.get(key) is None:
-                missed.append(key)
-        if missed:
-            raise KeyError(f'Key(s) {missed} were not found in {self.save_abspath}')
-
-        # hack
-        self.query_handler.close()
-        self.handler_mutex = False
-        self.delete(keys)
+        keys = check_keys_exist(keys, self.query_handler.header.keys(), self.save_abspath, self.logger)
```
Comment (JoanFM) [inline jina/executors/indexers/keyvalue.py:82]: if check does return something the name seems wrong. Also try to keep variables inmutable as much as possible

File: jina/executors/indexers/keyvalue.py (hunk 1 above)
Comment (cristianmtr) [reply jina/executors/indexers/keyvalue.py:88]: > the interfaces should never be changed, pasa executor parameter if needed but do not change interface or assume things will be passed

I do not EXPECT it to be there. That's why it's an optional `kwargs`. Again, just a small optimization

Comment (JoanFM) [reply jina/executors/indexers/keyvalue.py:88]: in which contrxt is this optimixzation and why does it only works sometimes?

**Comment by cristianmtr** [comment]
Not sure why I can't reply to this comment: https://github.com/jina-ai/jina/pull/1628#discussion_r553921838 
but @JoanFM here's the reply

When we call an .update, the keys will be computed for the update anyway. The update will then call the delete with only the filtered out keys.

If we don't do it this way, the .delete method will have to recalculate and filter out the missing keys.

When `.delete` is called on its own (not part of an update), then the optimization will not happen.

File: jina/executors/indexers/keyvalue.py (hunk 1 above)
Comment (cristianmtr) [reply jina/executors/indexers/keyvalue.py:88]: When we call an .update, the keys will be computed for the update anyway. The update will then call the delete with only the filtered out keys.

If we don't do it this way, the .delete method will have to recalculate and filter out the missing keys

File: jina/executors/indexers/keyvalue.py (hunk 3 above)
Comment (cristianmtr) [reply jina/executors/indexers/keyvalue.py]: Yes, good point, renamed

File: jina/helper.py [hunk 4]
```diff
@@ -693,12 +696,16 @@ def run(self):
...
+def filter_keys_nonexistent(keys: Iterator, keys_existing: Iterator, check_path: str, logger: 'JinaLogger'):
+    keys = set(keys)
+    missed = set()
+    for key in keys:
+        if key not in keys_existing:
+            missed.add(key)
+    if missed:
+        logger.warning(f'Key(s) {missed} were not found in {check_path}. Continuing anyway...')
+        keys -= missed
+    return keys
```
Comment (florian-hoenicke) [inline jina/helper.py:708]: `keys` could be a generator instead. Therefore we wouldn't have to load all keys in memory if used somewhere else. In our case we have the keys in memory anyways.

File: jina/helper.py [hunk 5]
```diff
@@ -693,12 +696,16 @@ def run(self):
         return asyncio.run(func(*args, **kwargs))
 
 
-def check_keys_exist(keys_to_check, existing_keys):
-    missed = []
-    for k in keys_to_check:
-        if k not in existing_keys:
-            missed.append(k)
-    return missed
+def filter_keys_nonexistent(keys: Iterator, keys_existing: Iterator, check_path: str, logger: 'JinaLogger'):
```
Comment (florian-hoenicke) [inline jina/helper.py:699]: `keys_existing` should not be type of Iterator but maybe Set.

File: jina/helper.py (hunk 4 above)
Comment (cristianmtr) [reply jina/helper.py:708]: Well they are already loaded in memory, for performing the calculation above

File: jina/helper.py (hunk 5 above)
Comment (cristianmtr) [reply jina/helper.py:699]: It depends on how the indexer stores the keys. I will cast them to `set` to be sure during the calculation

Comment (JoanFM) [reply jina/helper.py:699]: if this is to be used from different points then have a private _delete with an extra argument and have it called from update and delete method with the proper value.

Comment (JoanFM) [reply jina/helper.py:699]: Sorry is not in this discussion but innthe kwargs one

File: jina/executors/indexers/keyvalue.py (hunk 1 above)
Comment (cristianmtr) [reply jina/executors/indexers/keyvalue.py]: Response to your suggestion about `_delete`

Like so ? https://github.com/jina-ai/jina/blob/3d8d678a5591861aae1608589f4a08cec7306109/jina/executors/indexers/keyvalue.py#L82

Comment (JoanFM) [reply jina/executors/indexers/keyvalue.py]: yes more or less

File: jina/executors/indexers/cache.py [hunk 6]
```diff
@@ -99,9 +99,7 @@ def update(self, keys: Iterator['UniqueId'], values: Iterator[any], *args, **kwa
         """
         :param keys: list of Document.id
         :param values: list of either `id` or `content_hash` of :class:`Document"""
-        missed = check_keys_exist(keys, self.query_handler.ids)
-        if missed:
-            raise KeyError(f'Keys {missed} were not found in {self.index_abspath}. No operation performed...')
+        keys = filter_keys_nonexistent(set(keys), set(self.query_handler.ids), self.save_abspath, self.logger)
```
Comment (JoanFM) [inline jina/executors/indexers/cache.py:102]: can we call it `existent_keys`?

File: jina/executors/indexers/cache.py [hunk 7]
```diff
@@ -113,9 +111,7 @@ def delete(self, keys: Iterator['UniqueId'], *args, **kwargs):
         """
         :param keys: list of Document.id
         """
-        missed = check_keys_exist(keys, self.query_handler.ids)
-        if missed:
-            raise KeyError(f'Keys {missed} were not found in {self.index_abspath}. No operation performed...')
+        keys = filter_keys_nonexistent(set(keys), set(self.query_handler.ids), self.save_abspath, self.logger)
```
Comment (JoanFM) [inline jina/executors/indexers/cache.py:114]: same here?

File: jina/executors/indexers/vector.py [hunk 8]
```diff
@@ -137,15 +137,26 @@ def _check_keys(self, keys: Sequence[int]) -> None:
             raise KeyError(f'Key(s) {missed} were not found in {self.save_abspath}')
 
     def update(self, keys: Sequence[int], values: Sequence[bytes], *args, **kwargs) -> None:
-        self.delete(keys)
-        self.add(np.array(keys), np.array(values))
+        keys = filter_keys_nonexistent(set(keys), set(self.ext2int_id.keys()), self.save_abspath, self.logger)
```
Comment (JoanFM) [inline jina/executors/indexers/vector.py:140]: same here?

File: jina/executors/indexers/vector.py [hunk 9]
```diff
@@ -137,15 +137,26 @@ def _check_keys(self, keys: Sequence[int]) -> None:
...
+            self.delete(keys, **{'keys_precomputed': True})
+            self.add(np.array(keys), np.array(values))
 
     def delete(self, keys: Sequence[int], *args, **kwargs) -> None:
-        self._check_keys(keys)
-        for key in keys:
-            # mark as `False` in mask
-            self.valid_indices[self.ext2int_id[key]] = False
-            self._size -= 1
+        if kwargs.get('keys_precomputed') is not True:
```
Comment (JoanFM) [inline jina/executors/indexers/vector.py:149]: the idea would be to have in this class a 

```
def _delete(self, keys, keys_precomputed: bool = False)
```
that is called from different parts

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/executors/indexers/vector.py (hunk 9 above)
Comment (cristianmtr) [reply jina/executors/indexers/vector.py:149]: We don't even need the `precomputed_keys` arg if we opt for a private `_delete`. The public methods would simply handle the keys, since they are always called first. THis is the way I've implemented it now

File: .gitignore [hunk 10]
```diff
@@ -135,5 +135,6 @@ shell/jina-wizard.sh
 .idea
 
 # tests binaries
-/tests/**/*.bin
-*.bin
+**/**/*.bin.ids
```
Comment (JoanFM) [inline .gitignore:138]: this being needed is a smell that we are not properly using `tmpdir` in tests, where does this happen?

Comment (cristianmtr) [reply .gitignore:138]: the DocIDCache doesn't store the bin files properly when initiated on its own (not part of a flow)

Comment (JoanFM) [reply .gitignore:138]: why not? are u passing `test_metas`?

File: jina/executors/indexers/vector.py (hunk 9 above)
Comment (cristianmtr) [reply jina/executors/indexers/vector.py]: Also done for the numpyindexer now

File: tests/unit/executors/indexers/test_numpyindexer.py [hunk 11]
```diff
@@ -271,8 +271,7 @@ def test_numpy_update_delete(compress_level, test_metas):
     with BaseIndexer.load(save_abspath) as indexer:
         assert isinstance(indexer, NumpyIndexer)
         # this will fail cause the key doesn't exist
```
Comment (maximilianwerk) [inline tests/unit/executors/indexers/test_numpyindexer.py:273]: Remove/update comment

File: jina/helper.py [hunk 12]
```diff
@@ -693,12 +697,16 @@ def run(self):
         return asyncio.run(func(*args, **kwargs))
 
 
-def check_keys_exist(keys_to_check, existing_keys):
-    missed = []
-    for k in keys_to_check:
-        if k not in existing_keys:
-            missed.append(k)
-    return missed
+def filter_nonexistent_keys(keys: Iterator, existent_keys: Iterator, check_path: str, logger: 'JinaLogger'):
```
Comment (maximilianwerk) [inline jina/helper.py:700]: I would prefer having this method as a part of a base class of all the using classes. Might be `BaseIndexer`? Having the logger as an argument is a smell in my opinion.

**Review by maximilianwerk — COMMENTED** [review]
Just organizational/cleanup comments.

File: .gitignore (hunk 10 above)
Comment (cristianmtr) [reply .gitignore:138]: Hm I don't really remember. I will have a look at it in the upcoming PR where I'm adding the drivers for delete and update to DocCache

File: tests/unit/executors/indexers/test_numpyindexer.py (hunk 11 above)
Comment (cristianmtr) [reply tests/unit/executors/indexers/test_numpyindexer.py]: Yep, done

File: jina/helper.py (hunk 12 above)
Comment (JoanFM) [reply jina/helper.py:700]: I agree

Comment (cristianmtr) [reply jina/helper.py]: Ah good point. Done

File: .gitignore (hunk 10 above)
Comment (cristianmtr) [reply .gitignore]: Nvm, found the problem and fixed it

File: jina/executors/indexers/__init__.py [hunk 13]
```diff
@@ -162,6 +162,18 @@ def flush(self):
         except:
             pass
 
+    def filter_nonexistent_keys(self, keys: Iterator, existent_keys: Iterator, check_path: str):
```
Comment (JoanFM) [inline jina/executors/indexers/__init__.py:165]: can u make it "private"?

File: jina/helper.py [hunk 14]
```diff
@@ -21,6 +21,9 @@
 
 import numpy as np
 
+if False:
```
Comment (JoanFM) [inline jina/helper.py:24]: why is this needed'

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/helper.py (hunk 14 above)
Comment (cristianmtr) [reply jina/helper.py]: Ah it was from when the `filter_keys` was here. Removing the import left a `pass` behind it seems

File: jina/executors/indexers/__init__.py (hunk 13 above)
Comment (cristianmtr) [reply jina/executors/indexers/__init__.py]: done

**Review by JoanFM — APPROVED** [review]
